zng/docs/zson.md

```diff
@@ -0,0 +1,628 @@
+encoded with a null value.  If an instance of a record value omits a value
+by dropping the field altogether rather than using a null, then that record
+value corresponds to a different record type that elides the field in question.
+
+An _array type value_ has the form:
+```
+[ <type> ]
+```
+
+A _set type value_ has the form:
+```
+|[ <type> ]|
+```
+
+A _map type value_ has the form:
+```
+|{ <key-type>, <value-type> }|
+```
+where `<key-type>` is any type value of the keys and `<value-type>` is any
+type value of the values.
+
+A _union type value_ has the form:
+```
+( <type>, <type>, ... )
+```
+where there are at least two types in the list.
+
+An _enum type value_ has two forms.  The simple form is:
+```
+< <identifier>, <identifier>, ... >
+```
+and the complex form is:
+```
+< <identifier>:<value>, <identifier>=<value>, ... >
+```
+In the same form, the underlying enum value is equal to its positional
+index in the list of identifiers as an uint64 type.
+
+## Discussion and Examples
+
+> TBD: Add a range of helpful examples and also some that cover the more obscure corner cases.
+> Example of pico-second timestamp.
+
+### zeek
+```
+{
+    _path: "conn",
+    ts: 2018-03-24T17:15:20.600725Z,
+    uid: "C1zOivgBT6dBmknqk" (bstring),
+    id: {
+        orig_h: 10.47.1.152,
+        orig_p: 49562 (uint16) (=port),
+        resp_h: 23.217.103.245,
+        resp_p: 80 (port)
+    } (=24),
+    proto: "tcp" (=zenum),
+    service: null (bstring),
+    duration: 9.698493s (duration),
+    orig_bytes: 0 (uint64),
+    resp_bytes: 90453565 (uint64),
+    conn_state: "SF" (bstring),
+    local_orig: null (bool),
+    local_resp: null (bool),
+    missed_bytes: 0 (uint64),
+    history: "^dtAttttFf" (bstring),
+    orig_pkts: 57490 (uint64),
+    orig_ip_bytes: 2358856 (uint64),
+    resp_pkts: 123713 (uint64),
+    resp_ip_bytes: 185470730 (uint64),
+    tunnel_parents: null (|[bstring]| (=25))
+} (=26)
+{
+    _path: "conn",
+    ts: 2018-03-24T17:15:20.605945Z,
+    uid: "CayJxr1WvNLdJ6L9B4",
+    id: {
+        orig_h: 10.128.0.207,
+        orig_p: 8,
+        resp_h: 10.47.23.178,
+        resp_p: 0
+    },
+    proto: "icmp",
+    service: null,
+    duration: 4µs,
+    orig_bytes: 0,
+    resp_bytes: 0,
+    conn_state: "OTH",
+    local_orig: null,
+    local_resp: null,
+    missed_bytes: 0,
+    history: null,
+    orig_pkts: 2,
+    orig_ip_bytes: 56,
+    resp_pkts: 0,
+    resp_ip_bytes: 0,
+    tunnel_parents: null
+} (26)
+```
+
+### enum
+
+```
+{ rank: A (<2,3,4,5,6,7,8,9,10,Jack,Queen,King,Ace>), suit: H (<Hearts,Diamonds,Spaces,Clubs>) } (=card)
+```
+is the same as
+```
+{ rank: A, suit:H } ({rank: <2,3,4,5,6,7,8,9,10,Jack,Queen,King,Ace>, suit:<Hearts,Diamonds,Spaces,Clubs>}) (=card)
+```
+
+### union
+
+```
+{ u: 12 (int32, string, {a:int8,b:ip} (=27)), |[27]| } (=28) (=ex_union)
+{ u: "foo" } (=ex_union)
+```
+is the same as
+```
+{ u: 12 (int32, string, {a:int8,b:ip} (=27)), |[27]| } (=28) (=rec_with_union)
```

Comment:
probably missing something but I'm not understanding  `|[27]|` in:

`{ u: 12 (int32, string, {a:int8,b:ip} (=27)), |[27]| } (=28) (=ex_union)`.

Looks like a set of `{a:int8,b:ip}`, but should't there be a field name before it?

Reply:
Good catch.  I had a typo here and will fix... the set of records is supposed to be inside the union type.  In a future PR, I will add more examples and add discussion of the examples.